File: LeetCode/039.combination-sum/combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>>res = new ArrayList<>();
        Arrays.sort(candidates);
        solver(res, candidates, new ArrayList<>(), target, 0);
        return res;
    }

    public void solver(List<List<Integer>>res, int[] candidates, List<Integer>tmp, int target, int stIndex){
        if (target < 0) return;
        else if (target == 0)    res.add(new ArrayList<>(tmp));
        else 
        for (int i = stIndex; i < candidates.length; i++){
            tmp.add(candidates[i]);
            if (target - candidates[i] < 0) break;
            solver(res, candidates, new ArrayList<>(tmp), target - candidates[i], i);
            tmp.remove(tmp.size() - 1);
        }
    }
}

File: LeetCode/039.combination-sum/combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>>res;
        vector<int>tmp;
        solver(res, candidates, tmp, target, 0);
        return res;
    }

    void solver(vector<vector<int>>& res, vector<int>& candidates, vector<int>tmp, int target, int stIndex){
        if (target < 0) return;
        if (target == 0)    res.push_back(tmp);
        for (int i = stIndex; i < candidates.size(); i++){
            tmp.push_back(candidates[i]);
            solver(res, candidates, tmp, target - candidates[i], i);
            tmp.pop_back();
        }
    }
};